File: naru_agent/session/redis_store.py
from __future__ import annotations

import json
import logging
from typing import Any

from naru_agent.session.base import BaseSessionStore

logger = logging.getLogger(__name__)


class RedisSessionStore(BaseSessionStore):
    """Redis-backed session store for multi-instance deployments.

    Requires redis[asyncio]>=5.0:
        pip install "naru_agent[redis]"

    Usage:
        import redis.asyncio as aioredis
        client = aioredis.from_url("redis://localhost")
        store = RedisSessionStore(client, ttl=3600)
    """

    def __init__(self, client: Any, ttl: int | None = None, prefix: str = "naru:session:") -> None:
        self._client = client
        self._ttl = ttl
        self._prefix = prefix

    def _key(self, session_id: str) -> str:
        return f"{self._prefix}{session_id}"

    async def get(self, session_id: str) -> list[dict] | None:
        data = await self._client.get(self._key(session_id))
        if data is None:
            return None
        try:
            return json.loads(data)
        except (json.JSONDecodeError, TypeError) as e:
            logger.error(
                "Corrupted session data for '%s', discarding: %s", session_id, e,
            )
            return None

    async def save(self, session_id: str, history: list[dict]) -> None:
        key = self._key(session_id)
        value = json.dumps(history, ensure_ascii=False)
        if self._ttl is not None:
            await self._client.setex(key, self._ttl, value)
        else:
            await self._client.set(key, value)

    async def delete(self, session_id: str) -> None:
        await self._client.delete(self._key(session_id))

File: js/src/session/redis-store.ts
import type { ModelMessage } from "ai";
import type { BaseSessionStore } from "./base.js";

/**
 * Redis-backed session store. Requires `ioredis` as optional dependency.
 */
export class RedisSessionStore implements BaseSessionStore {
  private redis: import("ioredis").default;
  private prefix: string;
  private ttlSeconds: number;

  constructor(
    redis: import("ioredis").default,
    prefix = "naru:session:",
    ttlSeconds = 86400,
  ) {
    this.redis = redis;
    this.prefix = prefix;
    this.ttlSeconds = ttlSeconds;
  }

  private key(sessionId: string): string {
    return `${this.prefix}${sessionId}`;
  }

  async get(sessionId: string): Promise<ModelMessage[] | null> {
    const data = await this.redis.get(this.key(sessionId));
    if (!data) return null;
    return JSON.parse(data) as ModelMessage[];
  }

  async save(sessionId: string, history: ModelMessage[]): Promise<void> {
    await this.redis.set(
      this.key(sessionId),
      JSON.stringify(history),
      "EX",
      this.ttlSeconds,
    );
  }

  async delete(sessionId: string): Promise<void> {
    await this.redis.del(this.key(sessionId));
  }
}
